### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [274]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# Save file path to variable "pymoli_path"
pymoli_path = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame "pur_data"
pur_data = pd.read_csv(pymoli_path)
pur_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [275]:
# Slim pur_data to only contain columns of "SN", "Age", and "Gender" and save it as "player_info"
player_info = pur_data[["SN", "Age", "Gender"]]

# Drop duplicated rows in "player_info" 
player_info = player_info.drop_duplicates()

# Create a DataFrame called "total_player" to store the counts of "SN" in "player_info"
total_player = pd.DataFrame(
    {"Total Players": [player_info["SN"].count()] })

total_player

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [317]:
# Create a DataFrame to analyze purchase behaviors  
pur_glance = pd.DataFrame(
    {"Number of Unique Items": [len(pur_data["Item ID"].unique())],
     "Average Price": [pur_data["Price"].mean()],
     "Number of Purchases": [pur_data["Price"].count()],
     "Total Revenue": [pur_data["Price"].sum()]
    })

# Format "Average Price" and "Total Revenue" columns, respectively
pur_glance["Average Price"] = pur_glance["Average Price"].map("${:.2f}".format)
pur_glance["Total Revenue"] = pur_glance["Total Revenue"].map("${:,}".format)

pur_glance

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     183         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [277]:
# Save counts of each gender from "player_info" in a Series called "gender_count"
gender_count = player_info["Gender"].value_counts()


In [278]:
# Calculate and save player's count of  gender  percentage in a Series called "gender_pct"

gender_pct = player_info["Gender"].value_counts() * 100/player_info["SN"].count()

In [279]:
# Create a dataframe
gender_info = pd.DataFrame({"Total Count": gender_count,
    "Percentage of Players": gender_pct})


In [280]:
# Format value in "Percentage of Players" column
gender_info["Percentage of Players"] = gender_info["Percentage of Players"].map("{:.2f}".format)

gender_info

Total Count Percentage of Players
Male                           484                 84.03
Female                          81                 14.06
Other / Non-Disclosed           11                  1.91


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [281]:
# Use GroupBy in order to separate the original "pur_data" into fields according to "Gender"
grouped_gender_pur = pur_data.groupby(["Gender"], as_index = True)

# Create a new DataFrame called "pur_gender"
# Calculate and save "Purchase Count", "Average Purchase Price", "Total Purchase Value", and "Avg Total Purchase per Person" in "pur_gender"
pur_gender = pd.DataFrame(
    {"Purchase Count": grouped_gender_pur["Purchase ID"].count(),
     "Average Purchase Price": grouped_gender_pur["Price"].mean(),
     "Total Purchase Value": round(grouped_gender_pur["Price"].sum(), 2),
     "Avg Total Purchase per Person": grouped_gender_pur["Price"].sum() / gender_info["Total Count"]
    })

# Format "Average Purchase Price", "Total Purchase Value" and "Avg Total Purchase per Person" columns, respectively
pur_gender["Average Purchase Price"] = pur_gender["Average Purchase Price"].map("${:.2f}".format)
pur_gender["Total Purchase Value"] = pur_gender["Total Purchase Value"].map("${:,}".format)
pur_gender["Avg Total Purchase per Person"] = pur_gender["Avg Total Purchase per Person"].map("${:.2f}".format)


pur_gender

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [282]:
# Set up the bins for "player_info"
# Assume people with an age of 100+ won't be interested in gaming ... 
bins=[0,9,14,19,24,29,34,39,50]

# Create the names for binned categories
age_range = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Slice the data by binning "Age" column of "player_info" and place the data series into a new column called "Age Group"  
player_info["Age Group"] = pd.cut(player_info["Age"], bins,labels=age_range)
pur_data["Age Group"] = pd.cut(pur_data["Age"], bins,labels=age_range)
player_info.head()

SN  Age Gender Age Group
0        Lisim78   20   Male     20-24
1    Lisovynya38   40   Male       40+
2     Ithergue48   24   Male     20-24
3  Chamassasya86   24   Male     20-24
4      Iskosia90   23   Male     20-24

In [283]:
# Sort "player_info" based on "Age Group" column
player_info = player_info.sort_values("Age Group")

# Calculate the numbers by age group
age_count = player_info["Age Group"].value_counts()

# Calculate percentages of player's age group 
age_pct = player_info["Age Group"].value_counts() * 100/ len(player_info)


In [284]:

# Create a new DataFrame called "age_info"
age_info = pd.DataFrame({"Total Count": age_count, "Percentage of Players": age_pct})

# Format "Percentage of Players"
age_info["Percentage of Players"] = age_info["Percentage of Players"].map("{:.2f}".format)

# Sort rows by index
age_info.sort_index(axis=0,  ascending=True)


Total Count Percentage of Players
<10             17                  2.95
10-14           22                  3.82
15-19          107                 18.58
20-24          258                 44.79
25-29           77                 13.37
30-34           52                  9.03
35-39           31                  5.38
40+             12                  2.08

## Purchasing Analysis (Age)

* Bin the pur_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [285]:
# Calculate the purchase value by age group
Total_value_per_group = pur_data.groupby('Age Group').Price.sum().values

# Calculate the average purchase value by age group and  save the values as a list
average_purch=pur_data.groupby('Age Group').Price.mean().values

# Calculate the purchase times by age group and save the values as a list
purch_count = pur_data.groupby('Age Group').Price.count().values                    

# Calculate the Average total purchase per person and save the values as a list
# Average total purchase per person = total purchase value in each age group/ total number of player in each age group
avg_purch_per_person=pur_data.groupby('Age Group').Price.sum().values/age_count.sort_index(axis=0,  ascending=True).values


In [286]:
# Convert all data to a dataframe
raw_data_info = {'Purchase Count':purch_count,'Average Purchase Price':average_purch,
                  'Total Purchase Value':Total_value_per_group,
                 'Avg Total Purchase per Person':avg_purch_per_person}

df_columns=['Purchase Count','Average Purchase Price','Total Purchase Value','Avg Total Purchase per Person']
df_index=list(age_count.sort_index(axis=0,  ascending=True).keys())
purchase_age_df = pd.DataFrame(raw_data_info,columns=df_columns,index=df_index)

In [287]:
# Format the value in 'Average Purchase Price','Total Purchase Value',
# 'Avg Total Purchase per Person' columns
columns=['Average Purchase Price','Total Purchase Value','Avg Total Purchase per Person']
for c in columns:
    purchase_age_df[c] = purchase_age_df[c].map("${:.2f}".format)    
   

In [288]:
# Organize the rows, put the first row(index="<10") to the last one
purchase_age_df=purchase_age_df[1:].append(purchase_age_df[:1])
purchase_age_df

Purchase Count Average Purchase Price Total Purchase Value  \
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05             $1114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   
<10                23                  $3.35               $77.13   

      Avg Total Purchase per Person  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19  
<10                           $4.54

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [289]:
# Calculate the purchase times by SN group
Purchase_count = pd.DataFrame(pur_data.groupby('SN').count().Price)
# Rename the column
Purchase_count = Purchase_count.rename(columns={'Price':'Purchase Count'})


In [290]:
# Calculate the total purchase value by SN group
total_value = pd.DataFrame(pur_data.groupby('SN').sum().Price)
# Rename the column
total_value = total_value.rename(columns={'Price':'Total Purchase Value'})

In [291]:
# Combine two dataframe Purchase_count and total_value

top_spender_df = Purchase_count.join(total_value,how='left')


In [292]:
# Calculate the average purchase price
top_spender_df['Average Purchase Price']= top_spender_df['Total Purchase Value']/top_spender_df['Purchase Count']
top_spender_df.head()

Purchase Count  Total Purchase Value  Average Purchase Price
SN                                                                        
Adairialis76               1                  2.28                    2.28
Adastirin33                1                  4.48                    4.48
Aeda94                     1                  4.91                    4.91
Aela59                     1                  4.32                    4.32
Aelaria33                  1                  1.79                    1.79

In [293]:
# Sort the total purchase value column in descending order
# Slice the dataframe only keep the first five rows
top_spender_df = top_spender_df.sort_values(by=['Total Purchase Value'],ascending=False)[:5]

In [294]:
# Reorgnize the columns and format the dataframe 
top_spender_df = top_spender_df[['Purchase Count','Average Purchase Price','Total Purchase Value']]
top_spender_df['Total Purchase Value']=top_spender_df['Total Purchase Value'].map('${:.2f}'.format)
top_spender_df['Average Purchase Price']=top_spender_df['Average Purchase Price'].map('${:.2f}'.format)
top_spender_df

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [295]:
# Retrieve the Item ID, Item Name, and Item Price columns
Popular_Items_df=pur_data[['Item ID','Item Name','Price']]
Popular_Items_df.head()

Item ID                                  Item Name  Price
0      108  Extraction, Quickblade Of Trembling Hands   3.53
1      143                          Frenzied Scimitar   1.56
2       92                               Final Critic   4.88
3      100                                Blindscythe   3.27
4      131                                       Fury   1.44

In [296]:
# Calculate the total purchase value by 'Item ID','Item Name' columns
Popular_Items_df= Popular_Items_df.groupby(['Item ID','Item Name'])['Price'].value_counts()
# Convert to dataframe
Popular_Items_df= pd.DataFrame(Popular_Items_df)
# Remane column and convert the Price index to column
Popular_Items_df = Popular_Items_df.rename(columns={'Price':'Purchase Count'})
Popular_Items_df = Popular_Items_df.reset_index(level=['Price'])


In [297]:
# Remane column
Popular_Items_df = Popular_Items_df.rename(columns={'Price':'Item Price'})
Popular_Items_df.head()

Item Price  Purchase Count
Item ID Item Name                                     
0       Splinter                  1.28               4
1       Crucifer                  3.26               3
2       Verdict                   2.48               6
3       Phantomlight              2.49               6
4       Bloodlord's Fetish        1.70               5

In [329]:
# Check the puchase count value distribution
Popular_Items_df['Purchase Count'].describe()

count    183.000000
mean       4.262295
std        1.954721
min        1.000000
25%        3.000000
50%        4.000000
75%        5.000000
max       12.000000
Name: Purchase Count, dtype: float64

In [298]:
# Calculate the total purchase value
Popular_Items_df['Total Purchase Value']= (Popular_Items_df['Item Price'])* (Popular_Items_df['Purchase Count'])

In [299]:
# Format the value
Popular_Items_df['Item Price']=Popular_Items_df['Item Price'].map("${:.2f}".format)    
Popular_Items_df['Total Purchase Value']=Popular_Items_df['Total Purchase Value'].map("${:.2f}".format)

In [300]:
# Reorganize the columns name
Popular_Items_df= Popular_Items_df[['Purchase Count','Item Price','Total Purchase Value']]
# Sort the purchase count column in descending order
Popular_Items_df= Popular_Items_df. sort_values(by='Purchase Count', ascending=False)
Popular_Items_df.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [301]:
Popular_Items_df.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

In [302]:
# Convert the datatype from string to float in 'Total Purchase Value' column

Popular_Items_df['Total Purchase Value'] = Popular_Items_df['Total Purchase Value'].apply(lambda x:pd.to_numeric(x[1:]))
Popular_Items_df.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                      Total Purchase Value  
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm                 50.76  
145     Fiery Glass Crusader                                         41.22  
108     Extraction, Quickblade Of Trembling Hands                    31.77  
82      Nirvana                                                      44.10  
19      Pursuit, Cudgel of Necromancy                                 8.16

In [312]:
# Sort the dataframe by total purchase value in descending order
Profitable_Items =Popular_Items_df.sort_values(by='Total Purchase Value',ascending=False)


In [313]:
# Reorganize the columns
Profitable_Items = Profitable_Items[['Purchase Count','Item Price','Total Purchase Value']]
# Format the value
Profitable_Items['Total Purchase Value'] = Profitable_Items['Total Purchase Value'].map("${:.2f}".format)

In [314]:
# Check the change
Profitable_Items.head()

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
92      Final Critic                                               8   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80

## Conclusion:

    1. Polular items: We make 780 purchases on 183 items.There are 75% items sold more than three times. The highest player pruchase is 12 times(Item ID: 178, Item name: Oathbreaker, Last Hope of the Breaking Storm). 

    2. Most Popular and Profitable Items: 3 of the 5 displayed are in both lists, “Oathbreaker,Last Hope of the Breaking Storm” totaling to 12 players in total which alse has the highest total purchase value at $50.76. While “Pursuit, Cudgel of Necromancy”(Item ID:19), totaling 8 purchases and one of the top five most popular items, does not have a high total purchase value and is not one of the top five in total purchase value.

    3.Purchasing Analysis (Gender): Of the 576 players, the vast majority are male (84%), and the proportion of female players is only 14%. However, the data shows that the female($4.47) and other players ($4.56)show more average spend than male players($4.07)



